test/org/javarosa/xpath/expr/XPathFuncExprRandomizeTest.java

```diff
@@ -0,0 +1,175 @@
+        initializeNewInstance(formDef);
+        List<SelectChoice> choices1 = getSelectChoices(formDef, "/randomize/seededFruit2");
+
+        FormDef formDefAfterSerialization = serializeAndDeserializeForm(formDef);
+
+        initializeNewInstance(formDefAfterSerialization);
+        List<SelectChoice> choices2 = getSelectChoices(formDefAfterSerialization, "/randomize/seededFruit2");
+
+        assertTrue(nodesEqualInOrder(choices1, choices2));
+    }
+
+    private FormDef serializeAndDeserializeForm(FormDef formDef) throws IOException, DeserializationException {
+        // Initialize serialization
+        PrototypeManager.registerPrototypes(JavaRosaCoreModule.classNames);
+        PrototypeManager.registerPrototypes(CoreModelModule.classNames);
+        new XFormsModule().registerModule();
+
+        // Serialize form in a temp file
+        Path tempFile = createTempFile("javarosa", "test");
+        formDef.writeExternal(new DataOutputStream(newOutputStream(tempFile)));
+
+        // Create an empty FormDef and deserialize the form into it
+        FormDef deserializedFormDef = new FormDef();
+        deserializedFormDef.readExternal(
+            new DataInputStream(newInputStream(tempFile)),
+            PrototypeManager.getDefault()
+        );
+
+        delete(tempFile);
+        return deserializedFormDef;
+    }
+
+    private static void initializeNewInstance(FormDef formDef) {
+        formDef.initialize(true, new InstanceInitializationFactory());
+    }
+
+    private List<SelectChoice> getSelectChoices(FormDef formDef, String ref) {
+        FormIndex formIndex = getFormIndex(formDef, absoluteRef(ref));
+        FormEntryPrompt formEntryPrompt = new FormEntryPrompt(formDef, formIndex);
+        return formEntryPrompt.getSelectChoices();
+    }
+
+    private FormIndex getFormIndex(FormDef formDef, TreeReference ref) {
+        for (int localIndex = 0, lastIndex = formDef.getChildren().size(); localIndex < lastIndex; localIndex++)
+            if (formDef.getChild(localIndex).getBind().getReference().equals(ref))
+                return new FormIndex(localIndex, 0, ref);
+        throw new IllegalArgumentException("Reference " + ref + " not found");
+    }
+
+    private static boolean nodesEqualInOrder(List<SelectChoice> left, List<SelectChoice> right) {
+        if (left.size() != right.size())
+            return false;
+
+        for (int i = 0; i < left.size(); i++)
+            if (!left.get(i).getValue().equals(right.get(i).getValue()))
+                return false;
+
+        return true;
+    }
+
+    private static TreeReference absoluteRef(String path) {
+        TreeReference tr = new TreeReference();
+        tr.setRefLevel(REF_ABSOLUTE);
+        tr.setContext(CONTEXT_ABSOLUTE);
+        tr.setInstanceName(null);
+        Arrays.stream(path.split("/"))
```

Comment:
Oh, I didn’t realize we could use modern Java features in JavaRosa because it is linked into Collect. But the tests *don’t* run on Collect. Right?

Reply:
They don't so this should be 👌!

Reply:
Android Studio, with Collect, alerts us when we try to add a feature that is not present in the API versions we support. IDEA, with JavaRosa, does not. Stream is not available.

Reply:
I understand. So, even if we can use Java8, not all the API is supported by some of the Android API versions we need to support. I didn't know about all that :)

Reply:
I'm not totally sure whether this applies to dependencies! Would be good to check and document this. I /think/ what we're importing from Collect is bytecode so it may not matter. https://developer.android.com/studio/write/java8-support#supported_features doesn't make it clear to me either way but perhaps one of you will be able to know the naswer from there.